File: test/testVariableStepSizeFMU.cpp
// --------------------------------------------------------------
// Copyright (c) 2013, AIT Austrian Institute of Technology GmbH.
// All rights reserved. See file FMIPP_LICENSE for details.
// --------------------------------------------------------------

#include <stdlib.h>
#include <common/fmi_v1.0/fmiModelTypes.h>
#include <import/utility/include/VariableStepSizeFMU.h>

#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE testFMU
#include <boost/test/unit_test.hpp>

#define _USE_MATH_DEFINES
#include <cmath>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#ifndef WIN32
#include <signal.h>
void dummy_signal_handler( int ) {} // Dummy signal handler function.
#endif


BOOST_AUTO_TEST_CASE( test_fmu_load )
{
	std::string modelName( "sine_standalone" );
	VariableStepSizeFMU fmu( std::string( FMU_URI_PRE ) + modelName, modelName );
}


BOOST_AUTO_TEST_CASE( test_fmu_init )
{
#ifndef WIN32
	// Avoid that BOOST treats SIGCHLD signal as error.
	BOOST_REQUIRE( signal( SIGCHLD, dummy_signal_handler ) != SIG_ERR );
#endif

	std::string modelName( "sine_standalone" );
	VariableStepSizeFMU fmu( std::string( FMU_URI_PRE ) + modelName, modelName );

	std::string initRealInputNames[1] = { "omega" };
	double initRealInputVals[1] = { 0.1 * M_PI };

	const double startTime = 0.0;
	const double stepSize = 1.0; // NB: fixed step size enforced by FMU!

	int status = fmu.init( "test_sine", initRealInputNames, initRealInputVals, 1, startTime, stepSize );
	BOOST_REQUIRE_MESSAGE( 1 == status, "init(...) FAILED" );
}


BOOST_AUTO_TEST_CASE( test_fmu_getrealoutputs )
{
#ifndef WIN32
	// Avoid that BOOST treats SIGCHLD signal as error.
	BOOST_REQUIRE( signal( SIGCHLD, dummy_signal_handler ) != SIG_ERR );
#endif

	std::string modelName( "sine_standalone" );
	VariableStepSizeFMU fmu( std::string( FMU_URI_PRE ) + modelName, modelName );

	std::string initRealInputNames[1] = { "omega" };
	double initRealInputVals[1] = { 0.1 * M_PI };

	const double startTime = 0.0;
	const double stepSize = 1.0; // NB: fixed step size enforced by FMU!

	std::string realOutputNames[1] = { "x" };

	fmu.defineRealOutputs( realOutputNames, 1 );

	int status = fmu.init( "test_sine", initRealInputNames, initRealInputVals, 1, startTime, stepSize );
	BOOST_REQUIRE_MESSAGE( 1 == status, "init(...) FAILED" );

	double* result = fmu.getRealOutputs();
	BOOST_REQUIRE( result[0] == 0.0 );
}


BOOST_AUTO_TEST_CASE( test_fmu_run_simulation )
{
#ifndef WIN32
	// Avoid that BOOST treats SIGCHLD signal as error.
	BOOST_REQUIRE( signal( SIGCHLD, dummy_signal_handler ) != SIG_ERR );
#endif

	std::string modelName( "sine_standalone" );
	VariableStepSizeFMU fmu( std::string( FMU_URI_PRE ) + modelName, modelName );

	std::string initRealInputNames[1] = { "omega" };
	double initRealInputVals[1] = { 0.1 * M_PI };

	const double startTime = 0.0;
	const double stepSize = 1.0; // NB: fixed step size enforced by FMU!

	std::string realOutputNames[1] = { "x" };

	fmu.defineRealOutputs( realOutputNames, 1 );

	int status = fmu.init( "test_sine", initRealInputNames, initRealInputVals, 1, startTime, 2 * stepSize );
	BOOST_REQUIRE_MESSAGE( 1 == status, "init(...) FAILED" );

	const double stopTime = 5.0;
	double time = startTime;
	double* result;
	double reference;
	while ( time <= stopTime )
	{
		fmu.sync( time, time + stepSize );
		time += stepSize;

		result = fmu.getRealOutputs();
		reference = std::sin( 0.1 * M_PI * stepSize * time );

		BOOST_REQUIRE_MESSAGE( std::fabs( result[0] - reference ) < 1e-8,
				       "result mismatch: deltaResult = " << ( result[0] - reference ) );
	}
}

File: test/testVariableStepSizeFMU.py
# -------------------------------------------------------------------
# Copyright (c) 2013-2017, AIT Austrian Institute of Technology GmbH.
# All rights reserved. See file FMIPP_LICENSE for details.
# -------------------------------------------------------------------

import fmippim
import unittest
import math

class testVariableStepSizeFMU(unittest.TestCase):
  def test_fmi_1_0_load(self):
    model_name = 'sine_standalone'
    fmu = fmippim.VariableStepSizeFMU( FMU_URI_PRE + model_name, model_name )


  def test_fmi_1_0_init(self):
    model_name = 'sine_standalone';
    fmu = fmippim.VariableStepSizeFMU( FMU_URI_PRE + model_name, model_name )

    # construct string array for init parameter names
    init_vars = fmippim.new_string_array( 1 )
    fmippim.string_array_setitem( init_vars, 0, 'omega' )

    # construct double array for init parameter values
    init_vals = fmippim.new_double_array( 1 )
    fmippim.double_array_setitem( init_vals, 0, 0.1 * math.pi )

    start_time = 0.
    step_size = 1. # NB: fixed step size enforced by FMU!

    status = fmu.init( "test_sine", init_vars, init_vals, 1, start_time, step_size )
    self.assertEqual( status, 1 )


  def test_fmi_1_0_getrealoutputs(self):
    model_name = 'sine_standalone';
    fmu = fmippim.VariableStepSizeFMU( FMU_URI_PRE + model_name, model_name )

    # construct string array for init parameter names
    init_vars = fmippim.new_string_array( 1 )
    fmippim.string_array_setitem( init_vars, 0, 'omega' )

    # construct double array for init parameter values
    init_vals = fmippim.new_double_array( 1 )
    fmippim.double_array_setitem( init_vals, 0, 0.1 * math.pi )

    # construct string array with output names
    outputs = fmippim.new_string_array( 1 )
    fmippim.string_array_setitem( outputs, 0, 'x' )

    # define real output names
    fmu.defineRealOutputs( outputs, 1 );

    start_time = 0.
    step_size = 1. # NB: fixed step size enforced by FMU!

    status = fmu.init( "test_sine", init_vars, init_vals, 1, start_time, step_size )
    self.assertEqual( status, 1 )

    result = fmu.getRealOutputs()
    self.assertEqual( fmippim.double_array_getitem( result, 0 ), 0.0 ) # check value


  def test_fmi_1_0_runsimulation(self):
    model_name = 'sine_standalone';
    fmu = fmippim.VariableStepSizeFMU( FMU_URI_PRE + model_name, model_name )

    # construct string array for init parameter names
    init_vars = fmippim.new_string_array( 1 )
    fmippim.string_array_setitem( init_vars, 0, 'omega' )

    # construct double array for init parameter values
    init_vals = fmippim.new_double_array( 1 )
    fmippim.double_array_setitem( init_vals, 0, 0.1 * math.pi )

    # construct string array with output names
    outputs = fmippim.new_string_array( 1 )
    fmippim.string_array_setitem( outputs, 0, 'x' )

    # define real output names
    fmu.defineRealOutputs( outputs, 1 );

    start_time = 0.
    stop_time = 5.
    fmu_step_size = 1. # NB: fixed step size enforced by FMU!
    time = start_time

    status = fmu.init( "test_sine", init_vars, init_vals, 1, start_time, 2 * fmu_step_size )
    self.assertEqual( status, 1 )

    while ( time <= stop_time ):
      fmu.sync( time, time + fmu_step_size )
      time += fmu_step_size

      result = fmu.getRealOutputs()
      reference = math.sin( 0.1 * math.pi * fmu_step_size * math.floor( time / fmu_step_size ) )
      self.assertTrue( math.fabs( fmippim.double_array_getitem( result, 0 ) - reference ) < 1e-8 ) # check value



if __name__ == '__main__':
  import sys
  global EPS_TIME
  EPS_TIME = float( sys.argv.pop() )
  global FMU_URI_PRE
  FMU_URI_PRE = sys.argv.pop()
  unittest.main()
